osu.Game.Rulesets.Mania/Utils/SnapFinder.cs

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using osu.Framework.Utils;
+using osu.Game.Beatmaps.ControlPoints;
+using osu.Game.Rulesets.Mania.Beatmaps;
+using osu.Game.Rulesets.Objects;
+
+namespace osu.Game.Rulesets.Mania.Utils
+{
+    public class SnapFinder
+    {
+        private readonly ManiaBeatmap beatmap;
+
+        public SnapFinder(ManiaBeatmap beatmap)
+        {
+            this.beatmap = beatmap;
+        }
+
+        public int FindSnap(HitObject hitObject)
+        {
+            TimingControlPoint currentTimingPoint = beatmap.ControlPointInfo.TimingPointAt(hitObject.StartTime);
+            double startTime = currentTimingPoint.Time;
+            double secondsPerFourCounts = currentTimingPoint.BeatLength * 4;
```

Comment:
The tempo of a song is affected by the BPM value as well as type of beat that is used to measure the BPM. It seems osu measures the BPM in quavers (quarter notes). So for 100bpm, the duration of a quaver is the same whether it is a `SimpleQuadruple` or `SimpleTriple`.

What I'm doing here is converting from quavers per minute to semibreves (whole notes) per second. Then I can divide the whole note to find the snap

Reply:
Multiplying the beat length by 4 here does seem odd, as all we're looking for is the snap within a beat.

Right now we first take the modulus of 4 beats, then the modulus of a fraction of a beat. The former step is redundant, because the latter step is a smaller multiple of the beat length. I think it would be less confusing for readers of the code if we didn't do the former step at all, because it makes it seem like the structure of beats matter.

I'd suggest simply passing `hitObject.StartTime - currentTimingPoint.Time` to `almostDivisibleBy`, which yields the same result as far as I've tested.

Reply:
My latest commit already removed the * 4 multiplication because it's redundant, but yeah you're right, I don't need the modulus of 4 beats either! Looks like I got locked into thinking in whole notes while working on this